File: scripts/compare_states.py
#!/usr/bin/env python3
"""Compare two cuMES converged-state binaries (schema-v1 binary format).

Format: magic "CUMES001" (8) + int32 version + 2 ints (ns, mnmax), then 6
families of ns*mnmax doubles each, in the order rmncc, zmnsc, lmnsc, rmnss,
zmncs, lmncs (mode-major [m*ns + j]); the provenance trailer after the state
is ignored.

Usage: compare_states.py <state_a.bin> <state_b.bin>
Prints the max abs/rel difference per family and the per-mode max, skipping
the axis row (j=0), which is extrapolated and may differ between runs.
"""
import struct
import sys

FAMS = ("rmncc", "zmnsc", "lmnsc", "rmnss", "zmncs", "lmncs")
MAGIC = b"CUMES001"
HEADER = struct.Struct("<8siii")  # magic(8) version ns mnmax = 20 bytes


def load(path):
    with open(path, "rb") as f:
        head = f.read(HEADER.size)
        if len(head) != HEADER.size:
            raise SystemExit(f"error: {path} is not a v1 cumes state container")
        magic, version, ns, mnmax = HEADER.unpack(head)
        if magic != MAGIC or not (1 <= version <= 6) or ns < 1 or mnmax < 1:
            raise SystemExit(f"error: {path} is not a v1 cumes state container")
        n = ns * mnmax
        fams = {}
        for name in FAMS:
            fams[name] = struct.unpack(f"<{n}d", f.read(8 * n))
    return ns, mnmax, fams


def main():
    a_ns, a_mn, a = load(sys.argv[1])
    b_ns, b_mn, b = load(sys.argv[2])
    assert (a_ns, a_mn) == (b_ns, b_mn), f"size mismatch: {(a_ns, a_mn)} vs {(b_ns, b_mn)}"
    ns, mnmax = a_ns, a_mn
    print(f"ns={ns} mnmax={mnmax} (comparing interior j=1..{ns-1})")
    worst = 0.0
    for name in FAMS:
        amax = 0.0
        for m in range(mnmax):
            for j in range(1, ns):  # skip the extrapolated axis row
                i = m * ns + j
                d = abs(a[name][i] - b[name][i])
                scale = max(1.0, abs(b[name][i]))
                amax = max(amax, d / scale)
        worst = max(worst, amax)
        print(f"  {name:6s}: max rel diff (interior) = {amax:.3e}")
    print(f"worst over all families: {worst:.3e}")
    # Two runs stopped at the same residual floor (FSQR ~ 1e-13) can end at
    # state differences of ~1e-9 relative (the direct backend itself matches
    # vmecpp's wout at <=1.5e-9) — anything worse than 1e-8 indicates a
    # real transform discrepancy.
    ok = worst < 1e-8
    print("PASS" if ok else "FAIL")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/compare_states.cpp
// Compare two cuMES converged-state binary payloads.

#include "../include/clap.h"
#include "include/compare_common.hpp"

#include <algorithm>
#include <cerrno>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>

namespace {

struct CommandLine {
    std::string state_a;
    std::string state_b;
};

int parse_command_line(CommandLine& command, int argc, char** argv) {
    CLAP_BEGIN(CommandLine)
    CLAP_ADD_USAGE("STATE_A STATE_B")
    CLAP_ADD_DESCRIPTION(
        "Compare two cuMES state payloads, skipping the extrapolated axis row.")
    CLAP_REGISTER_ARG(state_a)
    CLAP_REGISTER_ARG(state_b)
    CLAP_END(CommandLine)
    try {
        CLAP<CommandLine>::parse_input(command, argc, argv);
    } catch (const std::exception& error) {
        std::cerr << error.what();
        return EINVAL;
    }
    return 0;
}

}  // namespace

int main(int argc, char** argv) {
    CommandLine command;
    if (const int status = parse_command_line(command, argc, argv); status) {
        return status;
    }

    try {
        const auto a = cumes::compare::read_state(command.state_a, false);
        const auto b = cumes::compare::read_state(command.state_b, false);
        if (a.ns != b.ns || a.mnmax != b.mnmax) {
            std::cerr << "error: size mismatch: (" << a.ns << ", " << a.mnmax
                      << ") vs (" << b.ns << ", " << b.mnmax << ")\n";
            return 2;
        }

        std::cout << "ns=" << a.ns << " mnmax=" << a.mnmax
                  << " (comparing interior j=1.." << a.ns - 1 << ")\n";
        double worst = 0.0;
        for (std::size_t family = 0;
             family < cumes::compare::FAMILY_NAMES.size(); ++family) {
            double family_worst = 0.0;
            for (std::int32_t mode = 0; mode < a.mnmax; ++mode) {
                for (std::int32_t surface = 1; surface < a.ns; ++surface) {
                    const auto index = static_cast<std::size_t>(mode) *
                                           static_cast<std::size_t>(a.ns) +
                                       static_cast<std::size_t>(surface);
                    family_worst = std::max(family_worst,
                                            cumes::compare::relative_difference(
                                                a.families[family][index],
                                                b.families[family][index]));
                }
            }
            worst = std::max(worst, family_worst);
            std::cout << "  " << std::left << std::setw(6)
                      << cumes::compare::FAMILY_NAMES[family] << std::right
                      << ": max rel diff (interior) = " << std::scientific
                      << std::setprecision(3) << family_worst << '\n';
        }
        std::cout << "worst over all families: " << std::scientific
                  << std::setprecision(3) << worst << '\n';
        constexpr double tolerance = 1.0e-8;
        const bool pass = worst < tolerance;
        std::cout << (pass ? "PASS\n" : "FAIL\n");
        return pass ? EXIT_SUCCESS : EXIT_FAILURE;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 2;
    }
}
